plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -506,10 +526,168 @@ export class PersonState {
 
             const deletePersonMessages = await this.db.deletePerson(otherPerson, client)
 
-            return [[...updatePersonMessages, ...distinctIdMessages, ...deletePersonMessages], person]
+            let personOverrideMessages: ProducerRecord[] = []
+            if (this.poEEmbraceJoin) {
+                personOverrideMessages = [await this.addPersonOverride(otherPerson, mergeInto, client)]
+            }
+
+            return [
+                [...personOverrideMessages, ...updatePersonMessages, ...distinctIdMessages, ...deletePersonMessages],
+                person,
+            ]
         })
     }
 
+    private async addPersonOverride(
+        oldPerson: Person,
+        overridePerson: Person,
+        client?: PoolClient
+    ): Promise<ProducerRecord> {
+        const mergedAt = DateTime.now()
+        const oldestEvent = overridePerson.created_at
+        /**
+            We'll need to do 4 updates:
+
+         1. Add the persons involved to the helper table (2 of them)
+         2. Add an override from oldPerson to override person
+         3. Update any entries that have oldPerson as the override person to now also point to the new override person
+
+            TODO: how do we want to be updating oldest_event?
+
+         I'm thinking: we write it if it's not there, but don't update it on conflicts (alternative update to the older date)
+         In the transitive one the same don't update or update to older data (if that's not too complex to do)
+         it's an optimization anyway and squash job might be updating those values anyway and it's just a hint, so ignoring if complex seems better
+         in any case it's important that we use the oldest_event we stored in postgres in CH too.
+         */
+        const oldPersonId = await this.addPersonOverrideHelper(oldPerson, client)
+        const overridePersonId = await this.addPersonOverrideHelper(overridePerson, client)
+
+        const {
+            rows: [{ version }],
+        } = await this.db.postgresQuery(
+            SQL`
+                INSERT INTO posthog_personoverride (
+                    team_id,
+                    old_person_id,
+                    override_person_id,
+                    oldest_event,
+                    version
+                ) VALUES (
+                    ${this.teamId},
+                    ${oldPersonId},
+                    ${overridePersonId},
+                    ${oldestEvent},
+                    1
+                )
+                RETURNING version;
+            `,
+            undefined,
+            'personOverride',
+            client
+        )
+
+        const { rows: transitiveUpdates } = await this.db.postgresQuery(
+            SQL`
+                WITH updated_ids AS (
+                    UPDATE
+                        posthog_personoverride
+                    SET
+                        override_person_id = ${overridePersonId}, version = version + 1
```

Comment:
This should be `version = COALESCE(version, 0)::numeric + 1` like we do for distinct_id moves. This is important for concurrent merges that might end up with two overrides with the same version - this would be a problem as both Kafka messages end up going to CH and the latter would randomly pick one of them to use (not what we want).

Reply:
I believe it's impossible for concurrent merges to run this query simultaneously as they both would lock the table, so only one would be allowed to proceed. The one that wasn't would just read the new updated value that the previous query set.

That being said, I don't see anything wrong with adding the `COALESCE` call (although it did made me question why we even allow `version` to be `null`).

Reply:
hmm... yeah I'm not a postgres expert, but consistency with pdi is probably good anyway